lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1314,19 +1317,48 @@ def _regenerate_alert_rules(self):
         self._generate_alert_rules_files(self.container)
         logger.debug("Saved alerts rules to disk")
 
-    def _endpoints(self) -> List[dict]:
-        """Return a list of Loki Push Api endpoints."""
-        return [{"url": self._url(unit_number=i)} for i in range(self._charm.app.planned_units())]
-
-    def _url(self, unit_number) -> str:
-        """Get the url for a given unit."""
-        return "http://{}-{}.{}-endpoints.{}.svc.cluster.local:{}/loki/api/v1/push".format(
-            self._charm.app.name,
-            unit_number,
-            self._charm.app.name,
-            self._charm.model.name,
-            self.port,
-        )
+    def update_endpoint(self, url: str = None, relation: Relation = None) -> None:
+        """Triggers programmatically the update of endpoint in unit relation data.
+
+        This method should be used when the charm relying on this library needs
+        to update the relation data in response to something occurring outside
+        of the `logging` relation lifecycle, e.g., in case of a
+        host address change because the charmed operator becomes connected to an
+        Ingress after the `logging` relation is established.
+
+        Args:
+            url: An optional url value to update relation data.
+            relation: An optional instance of `class:ops.model.Relation` to update.
+        """
+        if not relation:
+            if not self._charm.model.get_relation(self._relation_name):
+                return
+
+            relation = self._charm.model.get_relation(self._relation_name)
+
+        endpoint = self._endpoint(url or self._url)
+
+        relation.data[self._charm.unit].update({"endpoint": json.dumps(endpoint)})
+        logger.debug("Saved endpoint in unit relation data")
```

Comment:
The underlying idea with `update_endpoint` method is not also to provide a way to update endpoints data when a `logging_relation` is established (or changed) between Loki and a consumer charm but also to provide a way to update endpoints (from outside the lib) when an `ingress_relation` is established (or changed).
This way we are going to be able to forward ingress generated URLs to Loki's consumer.

The idea was:

1. An Ingress charm (let say Traefik), Loki and Loki's Consumer charm (Grafana agent) are deployed
2. A relation between Loki and Consumer charm is created
  a. Provider side (Loki) will update units relation data with Loki endpoints using `update_endpoint` method.
  b. Consumer side will get those endpoints using [`loki_endpoints`](https://github.com/canonical/loki-k8s-operator/blob/f9e4b216a062bcd8c72c4a036179ada9380e6f4f/lib/charms/loki_k8s/v0/loki_push_api.py#L1696) method and will update its config file with this data.
3. A relation between Loki and Traefik is created (`ingress-per-unit`).
  a. Traefik will create a new URL for each Loki unit. These URLs can be obtained using the Traefik's [`url` property](https://github.com/canonical/traefik-k8s-operator/blob/main/lib/charms/traefik_k8s/v0/ingress_per_unit.py#L739)
  b. We will get this new URL and call `update_endpoint` method from within the charm. This way we will forward this new endpoints to Loki's consumers since `update_endpoint` will change unit relation data.
  
We have a [work-in-progess PR here](https://github.com/canonical/loki-k8s-operator/pull/120) that needs to be updated once all our stabilisation PRs are merged.

Does this approach still make sense to you or you would prefer to modify it?

Reply:
Ah, I got it. Makes perfect sense to me now.